vizro-core/tests/unit/vizro/managers/test_data_manager.py

```diff
@@ -9,11 +10,26 @@
 import pytest
 from asserts import assert_frame_not_equal
 from flask_caching import Cache
+from freezegun import freeze_time
 from pandas.testing import assert_frame_equal
 from vizro import Vizro
 from vizro.managers import data_manager
 
 
+# TODO: Consider putting this inside the global 'tests/conftest.py' file with a scope of "session"?
+@pytest.fixture(autouse=True)
```

Comment:
Do you mind explaining how this speeds up the data_manager unit tests? I am a noob in this so it's not immediately clear from the code for me 😄

Reply:
So, the following lines of code ensure that every unit test in the `test_data_manager.py` have pinned system time to "2024-01-01 12:00:00". So the time is frozen inside the test (from the beginning to the end of the test), and time doesn't tick. 
```
@pytest.fixture(autouse=True)
def freeze_and_patch_time(monkeypatch):
    with freeze_time("2024-01-01 12:00:00", tick=False) as frozen_time:
        yield
```
It means that if you run `datetime.datetime.now()` or `time.time()` (or similar) inside the test, the returned value will be "2024-01-01 12:00:00". It's a crucial move as flask-caching library relies a lot on the system time (see the[issue](https://github.com/McK-Internal/vizro-internal/issues/957) description). So, if `data_frame` timeout is set like this `data_manager["data_1"].timeout=15`, the expiration time (calculated from the flask-caching) for the `data_1` will be 2024:01:01-12:00:15 now.

In the current tests implementation we use `time.sleep(N)` to ensure that data_frame cache expires. But, it means that the most of our unit tests will fail because time doesn't tick and cached data_frame will never expire. So, to ensure that time.sleep(N) moves the time for `N` seconds, I patched time.sleep function so `it moves the frozen time for N seconds`. To implement this I used the following code:
```
def advance_time(seconds):
            # Advance the frozen time by the given number of seconds
            frozen_time.move_to(frozen_time.time_to_freeze + datetime.timedelta(seconds=seconds))

        # Patch time.sleep to use advance_time instead of actual sleeping
        monkeypatch.setattr(time, "sleep", advance_time)
```

So, the speedup comes as a side product of the time.sleep patching as time.sleep doesn't really "sleep" for N seconds but it triggers the `advance_time` function that moves the frozen time for N seconds instantly.

Reply:
Thanks for the explanation - this helps a lot! 🚀